### PR TITLE
Refactor api package to improve pkg.go.dev experience.

### LINKED ISSUE #5863: API package should be godoc friendly
### What

Take a look at the example of the `autocert` package docs:

https://pkg.go.dev/golang.org/x/crypto/acme/autocert#pkg-overview

And compare to Teleport's API

https://pkg.go.dev/github.com/gravitational/teleport/api@v0.0.0-20210304221830-6d8f77815793/client

Add a couple of inline examples, and improve comments to make pkg.go.dev documentation more friendly to beginners.

### PR DESCRIPTION
**Pull request by Joerger**
This PR fixes several issues with the current iteration of the api documenation.

Changes:
 - godoc / pkg.go.dev improvements:
   - Add REAME to `client` package which loads at the top of the pkg.go.dev site with a documentation reference section.
   - Refactor a variety of useful comments in the `api/client` package.
   - Refactored api package to have more sensible sub packages. This structural improvement is very big if you compare the old godoc with the new godoc.
     -  Un-exported unneeded types like `ProfileCreds` which can be exported as `Credentials`.
     - Reorganized files to reduce size of client package. Moved profile, identityfile, and webclient into `api` subpackages.
   - `pkg.go.dev` can be run locally:
     - `git clone git@github.com:golang/pkgsite.git` 
     - `cd pkgsite && go run ./pkgsite/cmd/pkgsite/main.go [path-to]/gravitational/teleport/api`
     - go to http://localhost:8080/github.com/gravitational/teleport/api/client
 - Replace current [api documentation](https://goteleport.com/docs/api-reference/), with 4 new pages: API reference, introduction, getting started, and architecture.

closes #5863 #5466 #5860

### CODE REVIEW

File: api/client/README.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
+### Introduction
+
+The Teleport API Client provides secure access to the gRPC Auth API used by Teleport's CLIs ([tsh](https://goteleport.com/docs/cli-docs#tsh) and [tctl](https://goteleport.com/docs/cli-docs#tctl).
```
Comment (klizhentas) [inline api/client/README.md:3]: Let's remove secure and replace with mTLS (mutual TLS) what is more factual.

File: api/client/README.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+A fair amount of Teleport knowledge is expected before getting started. Make sure to familiarize yourself with the [admin manual](https://goteleport.com/docs/admin-guide/) and other [Teleport documentation](https://goteleport.com/docs/).
```
Comment (klizhentas) [inline api/client/README.md:5]: You can remove the first sentence without the second one loosing meaning.

File: api/client/README.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+#### Authentication and Authorization
+
+The client uses TLS certificates signed by a Teleport Auth server to authenticate its connection. 
+These TLS certificates must be associated with a Teleport user.
+
+Once authenticated, the client will be authorized to call client methods that the underlying user has
+permissions for.
+
+For example, the client user must have permission to perform `read` action on the `role` resource in order to call:
```
Comment (klizhentas) [inline api/client/README.md:17]: in order to create a role:

File: api/client/README.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+```go
+role, err := clt.GetRole(ctx, "role1")
+```
+
+See our [roles documentation](https://goteleport.com/docs/access-controls/reference/#roles) for more details
+on role based access control.
+
+#### Credentials
```
Comment (klizhentas) [inline api/client/README.md:25]: This goes to arch

File: api/client/README.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+| Ease of use | easy | easy | med | hard | 
+| Supports long lived certificates | not easily (configurable on server side) | yes | yes | yes | 
+| Supports proxy connections | yes | yes (6.1+) | no | no | 
+| Automatic Proxy Address discovery | yes | no | no | no | 
+| CLI used | tsh | tctl/tsh | tctl | - | 
+| Available in | 6.1+ | 6.0+ | 6.0+ | 6.0+ | 
+
+See the [Credentials type](https://pkg.go.dev/github.com/gravitational/teleport/api/client#Credentials) on pkg.go.dev for more information and examples for each.
+
+#### Additional resources
```
Comment (klizhentas) [inline api/client/README.md:58]: This is the next steps link in the getting started guide

File: api/client/README.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+The Teleport client is mainly documented through [pkg.go.dev](https://pkg.go.dev/github.com/gravitational/teleport/api/client). 
+
+For supplementary documentation about using Teleport, visit the main Teleport [docs](https://goteleport.com/docs/).
+
+### Getting started
+
+Before you begin:
+
+- Install [Go](https://golang.org/doc/install) 1.15+ and Setup Go Dev Environment
```
Comment (klizhentas) [inline api/client/README.md:68]: this is a template variable in the docs `(=teleport.golang=)`

File: api/client/README.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+- Set up Teleport with the [Getting Started Guide](https://goteleport.com/docs/getting-started/)
+
+Once you have the required rerequisites, simply set up a new [go module](https://golang.org/doc/tutorial/create-module) and import the `client` package:
```
Comment (klizhentas) [inline api/client/README.md:71]: prerequisites

File: api/client/README.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+Before you begin:
+
+- Install [Go](https://golang.org/doc/install) 1.15+ and Setup Go Dev Environment
+- Set up Teleport with the [Getting Started Guide](https://goteleport.com/docs/getting-started/)
+
+Once you have the required rerequisites, simply set up a new [go module](https://golang.org/doc/tutorial/create-module) and import the `client` package: 
+
+```bash
+$ go mod init client-demo
+$ go get github.com/gravitational.com/api/client
```
Comment (klizhentas) [inline api/client/README.md:75]: `go get github.com/gravitational/teleport/api/client`?

File: api/client/README.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+	"time"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/types"
+)
+
+func main() {
+	ctx := context.Background()
+
+	// Create a new client in your go file.
```
Comment (klizhentas) [inline api/client/README.md:166]: @Joerger great job. This is really easy to use.

**Review by klizhentas — CHANGES_REQUESTED** [review]
@Joerger @awly I only reviewed the docs

File: api/client/README.md [hunk 10]
```diff
@@ -0,0 +1,227 @@
+### Introduction
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/client/README.md:1]: Thank you for these resources, this should help organize it better. I do want to keep something in the README because it can add some nice flavor to the top of the `pkg.go.dev`. See [aws-sdk](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2) for example.

How about I move the bulk of the docs to `goteleport.com/docs` as suggested, and then add abbreviated sections with links to the docs in the README. That way whether a user starts from `pkg.go.dev` or the main docs, they'll find all the documentation with ease.

**Comment by Joerger** [comment]
@awly @klizhentas I'd recommend running pkg.go.dev locally to review that side of the documentation.

File: api/client/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (klizhentas) [reply api/client/README.md]: yes that works

File: api/client/README.md [hunk 11]
```diff
@@ -0,0 +1,10 @@
+This package is documented using a combination of [pkg.go.dev](https://pkg.go.dev/github.com/gravitational/teleport/api/client) and Teleport [Docs](https://goteleport.com/docs/).
+
+## Reference
+
+- [Introduction](https:/goteleport.com/docs/reference/api/introduction)
```
Comment (klizhentas) [inline api/client/README.md:5]: missing `/` here and below and missing trailing `/` as well

File: api/client/README.md [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+- [Getting Started](https:/goteleport.com/docs/reference/api/getting-started)
+- [Architecture](https:/goteleport.com/docs/reference/api/architecture)
+- [pkg.go.dev](https://pkg.go.dev/github.com/gravitational/teleport/api/client)
+  - [Client type](https://pkg.go.dev/github.com/gravitational/teleport/api/client#Client)
```
Comment (klizhentas) [inline api/client/README.md:9]: I'd remove type as it's not clear what it means in this context

File: docs/pages/reference/api.mdx [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+description: Reference for the Teleport API
+---
+
+# Reference
+
+- [Introduction](./api/introduction.mdx)
+- [Getting Started](./api/getting-started.mdx)
+- [Architecture](./api/architecture.mdx)
+- [pkg.go.dev](https://pkg.go.dev/github.com/gravitational/teleport/api/client)
+  - [Client type](https://pkg.go.dev/github.com/gravitational/teleport/api/client#Client)
```
Comment (klizhentas) [inline docs/pages/reference/api.mdx:12]: `Using client`
`Working with credentials`

File: docs/pages/reference/api/architecture.mdx [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+## Credentials
+
+The Teleport Go client uses Credentials in order to gather and hold TLS certificates, connect 
+to proxy servers over SSH, and perform some other actions.
+
+Credentials are created by using Credential loaders, which gather certificates and data 
+generated by [Teleport CLIs](../../cli-docs.mdx).
+
+Since there are several Credential loaders to choose from with distinct benefits, here's a quick breakdown:
```
Comment (klizhentas) [inline docs/pages/reference/api/architecture.mdx:56]: Such a well-written guidance, awesome.

File: docs/pages/reference/api/architecture.mdx [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+See the [Credentials type](https://pkg.go.dev/github.com/gravitational/teleport/api/client#Credentials)
+on pkg.go.dev for more information and examples for Credentials and Credential Loaders.
+
+## Client Connection
+
+The API client makes requests through an open connection to the Teleport Auth server.
+This connection can be made directly or remotely.
+
+Direct Auth server connections can only be made if the client is running locally to the server.
```
Comment (klizhentas) [inline docs/pages/reference/api/architecture.mdx:87]: Not quite sure what do you mean here, client can issue direct mTLS connections even if it's running on a different machine, but has a networking connection to the servers HTTPS address.

File: docs/pages/reference/api/architecture.mdx [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+Remote Auth server connections can be made through the server's reverse tunnel address using SSH certificates
```
Comment (klizhentas) [inline docs/pages/reference/api/architecture.mdx:89]: That's used when auth server is behind the proxy. It does not necessarily improve security, but is a good additional layer of network isolation for the auth servers.

File: docs/pages/reference/api/getting-started.mdx [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+
+In this getting started guide we will use the Teleport API Go Client to connect to the Teleport Auth server.
+
+Here are the steps we'll walk through:
+ 1. Create an API user
+ 2. Generate credentials
+ 3. Create a Go program to demo the client
+
+<Admonition type="warning">
+  This quickstart connects to the Teleport web proxy with an insecure connection. In production,
```
Comment (klizhentas) [inline docs/pages/reference/api/getting-started.mdx:16]: Let's always use secure connection in all guides. You can leave it to the user to use insecure mode if they need it. The danger that is that people will copy-paste the guide and will forget to remove the insecure flags.

File: docs/pages/reference/api/getting-started.mdx [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+func main() {
+	ctx := context.Background()
+
+	clt, err := client.New(ctx, client.Config{
+		Credentials: []client.Credentials{
+			client.LoadProfile("", ""),
+		},
+		// set to true if your Teleport web proxy doesn't have an HTTP/TLS
+		// certificate configured yet (never use this in production).
+		InsecureAddressDiscovery: true,
```
Comment (klizhentas) [inline docs/pages/reference/api/getting-started.mdx:83]: Let's set to false by default

File: docs/pages/reference/api/introduction.mdx [hunk 19]
```diff
@@ -0,0 +1,34 @@
+---
+title: Teleport API Introduction
+description: Introduction for the Teleport API
+---
+
+The Teleport Auth API provides a gRPC API for remotely interacting with a Teleport Auth server.
+
+We recently created a public [Go client](https://pkg.go.dev/github.com/gravitational/teleport/api/client)
```
Comment (klizhentas) [inline docs/pages/reference/api/introduction.mdx:8]: Let's remove `we recently`. This article will stay evergreen for a while, so this will become not relevant in a year.

File: docs/pages/reference/api/introduction.mdx [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+to programatically interact with the API. [tsh and tctl](../../cli-docs.mdx/) use the same API.
+
+## Go Client
+
+The Go client can be used to programatically interact with Teleport. For example:
```
Comment (klizhentas) [inline docs/pages/reference/api/introduction.mdx:13]: Here is what you can do with Teleport Go Client:

File: docs/pages/reference/api/introduction.mdx [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+ - Integrating with external tools, which we have already done
```
Comment (klizhentas) [inline docs/pages/reference/api/introduction.mdx:14]: such as Slack, Jira or Mattermost

**Review by klizhentas — CHANGES_REQUESTED** [review]
Almost there, let's do another pass and good to go!

File: api/client/credentials.go [hunk 22]
```diff
@@ -170,61 +194,59 @@ func (c *IdentityCreds) SSHClientConfig() (*ssh.ClientConfig, error) {
...
+//
+// Profile Credentials can be used to connect to an auth server directly
+// or through a reverse tunnel.
+//
+// Profile Credentials will automatically attempt to find your reverse
+// tunnel address and make a connection through it.
+//
+// A new profile can be generated with tsh.
+//  $ tsh login --user=api-user
+func LoadProfile(dir, name string) Credentials {
```
Comment (awly) [inline api/client/credentials.go:218]: Document the `dir` and `name` arguments

File: api/client/contextdialer.go [hunk 23]
```diff
@@ -34,23 +35,23 @@ type ContextDialer interface {
...
 type ContextDialerFunc func(ctx context.Context, network, addr string) (net.Conn, error)
 
 // DialContext is a function that dials to the specified address
 func (f ContextDialerFunc) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
 	return f(ctx, network, addr)
 }
 
-// NewDialer makes a new dialer.
+// NewDialer makes a new dialer to connect directly to an Auth server.
 func NewDialer(keepAlivePeriod, dialTimeout time.Duration) ContextDialer {
```
Comment (awly) [inline api/client/contextdialer.go:47]: How about exporting `NewDirectDialer` and `NewProxyDialer`?
`NewDialer` becomes `NewDirectDialer`.
`NewTunnelDialerWithAddressDiscovery` becomes `NewProxyDialer`.
`NewTunnelDialer` is unexported because users wouldn't ever need to use it (but correct me if I'm wrong)

File: api/client/doc_test.go [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+			Rules: []types.Rule{
+				types.NewRule(types.KindAccessRequest, []string{types.VerbList, types.VerbRead}),
+			},
+		},
+		Deny: types.RoleConditions{
+			NodeLabels: types.Labels{"*": []string{"*"}},
+		},
+	}
+
+	// There are helper functions for dealing with Teleport resources.
```
Comment (awly) [inline api/client/doc_test.go:52]:
```suggestion
	// There are helper functions for creating Teleport resources.
```

File: api/client/doc_test.go [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+	role, err := types.NewRole("role1", roleSpec)
+	if err != nil {
+		log.Fatalf("failed to get role: %v", err)
+	}
+
+	// Getters and setters can be used to alter specs.
+	role.SetLogins(types.Allow, []string{""})
```
Comment (awly) [inline api/client/doc_test.go:59]:
```suggestion
	role.SetLogins(types.Allow, []string{"root"})
```

File: api/client/doc_test.go [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+		// must be provided, except when using the ProfileCreds.
+		Addrs: []string{
+			// The Auth server address can be provided to connect locally.
+			"localhost:3025",
+			// The tunnel proxy address can be provided
+			// to connect to the Auth server over SSH.
+			"public_address:3024",
+			// The web proxy address can be provided to automatically
+			// find the tunnel proxy address and connect using it.
+			"public_address:3080",
```
Comment (awly) [inline api/client/doc_test.go:94]:
```suggestion
			"auth.example.com:3025",
			// The tunnel proxy address can be provided
			// to connect to the Auth server over SSH.
			"proxy.example.com:3024",
			// The web proxy address can be provided to automatically
			// find the tunnel proxy address and connect using it.
			"proxy.example.com:3080",
```

File: api/profile/profile.go [hunk 27]
```diff
@@ -216,11 +216,11 @@ func defaultProfilePath() string {
 	return filepath.Join(home, profileDir)
 }
 
-// ProfileFromDir reads the user (yaml) profile from a given directory. If
+// FromDir reads the user (yaml) profile from a given directory. If
```
Comment (awly) [inline api/profile/profile.go:219]:
```suggestion
// FromDir reads the user profile from a given directory. If
```

File: api/profile/profile.go [hunk 28]
```diff
@@ -284,34 +284,42 @@ func (p *Profile) saveToFile(filepath string) error {
...
-func (p *Profile) sshCertPath() string {
-	return filepath.Join(p.sshDir(), p.SiteName+constants.FileExtSSHCert)
+// KeyDir returns the path to the profile's ssh certificate.
+func (p *Profile) SSHCertPath() string {
+	return filepath.Join(p.SSHDir(), p.SiteName+constants.FileExtSSHCert)
 }
 
-func (p *Profile) sshCasPath() string {
+// KeyDir returns the path to the profile's ssh certificate authorities.
+func (p *Profile) SSHCasPath() string {
```
Comment (awly) [inline api/profile/profile.go:323]: nit: update the start of each comment to match the method name

File: api/profile/profile.go [hunk 29]
```diff
@@ -284,34 +284,42 @@ func (p *Profile) saveToFile(filepath string) error {
...
-func (p *Profile) userKeyDir() string {
-	return filepath.Join(p.keyDir(), p.Name())
+// KeyDir returns the path to the profile's key directory.
+func (p *Profile) UserKeyDir() string {
+	return filepath.Join(p.KeyDir(), p.Name())
 }
 
-func (p *Profile) keyPath() string {
-	return filepath.Join(p.userKeyDir(), p.Username)
+// KeyDir returns the path to the profile's PEM-encoded key.
```
Comment (awly) [inline api/profile/profile.go:297]:
```suggestion
// KeyDir returns the path to the profile's private key.
```

File: api/profile/profile.go [hunk 30]
```diff
@@ -284,34 +284,42 @@ func (p *Profile) saveToFile(filepath string) error {
...
-	return filepath.Join(p.userKeyDir(), p.Username+constants.FileExtTLSCert)
+// KeyDir returns the path to the profile's TLS certificate.
+func (p *Profile) TLSCertPath() string {
+	return filepath.Join(p.UserKeyDir(), p.Username+constants.FileExtTLSCert)
 }
 
-func (p *Profile) tlsCasPath() string {
-	return filepath.Join(p.userKeyDir(), constants.FileNameTLSCerts)
+// KeyDir returns the path to the profile's TLS certificate authorities.
+func (p *Profile) TLSCasPath() string {
```
Comment (awly) [inline api/profile/profile.go:308]:
```suggestion
func (p *Profile) TLSCAsPath() string {
```

File: api/profile/profile.go (hunk 28 above)
Comment (awly) [inline api/profile/profile.go:323]:
```suggestion
func (p *Profile) SSHCAsPath() string {
```

File: api/client/webclient/webclient.go [hunk 31]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package client
+package webclient
```
Comment (awly) [inline api/client/webclient/webclient.go:17]: Add package doc

File: api/identityfile/identityfile.go [hunk 32]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package client
+package identityfile
```
Comment (awly) [inline api/identityfile/identityfile.go:17]: Add package doc

File: api/profile/profile.go [hunk 33]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package client
+package profile
```
Comment (awly) [inline api/profile/profile.go:17]: Add package doc

File: api/client/client.go [hunk 34]
```diff
@@ -71,30 +77,22 @@ type Client struct {
...
-//     LoadIdentityFile("identity-file-path")
+// New will try to open a connection with all combinations of addresses and credentials.
+// The first successful connection to a server will be used, or an aggregated error will
+// be returned if all combinations fail.
 //
-// New will try to open a connection with all combinations of addresses,
-// server types, and credentials. The first successful connection to a server
-// will be used, or an aggregated error will be returned if all combinations fail.
+// At least one Credentials object and one address or dialer must be provided. If Profile
+// Credentials are used, a web proxy dialer will automatically be provided.
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/client/client.go:85]: Yes, I'll try to reword it to be more clear.

**Review by klizhentas — APPROVED** [review]
@Joerger looks good. Let's merge and backport to branch/v6 (future 6.2). No need to backport to branch/v6.1 (6.1) for now.

File: api/client/credentials.go [hunk 35]
```diff
@@ -17,84 +17,98 @@ limitations under the License.
...
 
-// LoadKeyPair is used to load credentials from files on disk.
-func LoadKeyPair(certFile string, keyFile string, caFile string) *KeyPairCreds {
-	return &KeyPairCreds{
+// LoadKeyPair is used to load Credentials from a certicate keypair on disk.
+//
+// KeyPair Credentials can only be used to connect directly to a local Teleport Auth server.
+//
+// New KeyPair files can be generated with tsh or tctl.
+//  $ tctl auth sign --format=tls --user=api-user --out=path/to/certs
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/client/credentials.go:85]: Is this possible? I don't see it in the CLI help or docs.

[the remark this answers is not in the document]
Comment (awly) [reply api/client/credentials.go:85]: Ah, I thought it would be `tsh login --format=tls --out=filename` but that `--format=tls` is not valid in `tsh` for some reason.
In that case, update the above sentence to not mention `tsh`

File: api/identityfile/identityfile.go [hunk 36]
```diff
@@ -14,7 +14,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package client
+// Package identityfile provides the IdentityFile struct and its methods.
```
Comment (awly) [inline api/identityfile/identityfile.go:17]: nit: prefer to describe the purpose of the package rather than its contents
```suggestion
// Package identityfile implements parsing and serialization of Teleport identity files.
```

File: api/client/webclient/webclient.go [hunk 37]
```diff
@@ -14,7 +14,9 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package client
+// Package webclient provides an implementation for the
+// Teleport web client Ping and Find http endpoints.
```
Comment (awly) [inline api/client/webclient/webclient.go:18]:
```suggestion
// Package webclient provides a client for the Teleport Proxy API endpoints.
```

File: api/profile/profile.go [hunk 38]
```diff
@@ -14,7 +14,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package client
+// Package profile provides the tsh Profile struct and its methods and functions.
```
Comment (awly) [inline api/profile/profile.go:17]:
```suggestion
// Package profile handles management of the Teleport profile directory (~/.tsh).
```

**Review by awly — APPROVED** [review]

**Comment by Joerger** [comment]
@russjones @r0mant bot.

**Review by russjones — APPROVED** [review]
